core/templates/pages/exploration-editor-page/improvements-tab/services/improvement-playthrough-modal.controller.ts

```diff
@@ -0,0 +1,175 @@
+angular.module('oppia').controller('ImprovementPlaythoughModalController', [
+  '$scope', '$uibModalInstance', 'AlertsService', 'LearnerActionRenderService',
+  'playthrough', 'playthroughIndex',
+  'ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS',
+  'MAX_UNRELATED_ACTIONS_PER_BLOCK',
+  function(
+      $scope, $uibModalInstance, AlertsService, LearnerActionRenderService,
+      playthrough, playthroughIndex,
+      ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS,
+      MAX_UNRELATED_ACTIONS_PER_BLOCK) {
+    let indexOfFirstDisplayedAction = null;
+    let actionIndexAtStartOfFinalBlock = null;
+
+    let init = function() {
+      $scope.playthroughIndex = playthroughIndex;
+
+      // This is the index of the first action in the range of actions
+      // displayed. Initially, only one action is displayed. When this
+      // index reaches 0, we've displayed all the learner actions.
+      indexOfFirstDisplayedAction = playthrough.actions.length - 1;
+
+      $scope.expandActionsToRender();
+
+      // Index to know where to start highlighting actions. This is
+      // set to the index of the first learner action in the final
+      // display block, and is never updated after that.
+      actionIndexAtStartOfFinalBlock = indexOfFirstDisplayedAction;
+
+      $scope.issueIsMultipleIncorrectSubmissions = false;
+      if (playthrough.issueType === (
+        ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS)) {
+        $scope.issueIsMultipleIncorrectSubmissions = true;
+      }
+    };
+
+    /**
+     * Identifies whether all learner actions have been displayed and
+     * no further expansion is possible.
+     * @returns {boolean}
+     */
+    $scope.canExpandActions = function() {
+      return indexOfFirstDisplayedAction !== 0;
+    };
+
+    /**
+     * Iterates through the learner actions and expands the set of
+     * actions to be displayed. If all learner actions have been
+     * handled, no more expansion is possible.
+     */
+    $scope.expandActionsToRender = function() {
+      let i;
+      let previousStateName = null;
+      if (indexOfFirstDisplayedAction < playthrough.actions.length - 1) {
+        let action = playthrough.actions[
+          indexOfFirstDisplayedAction];
+        previousStateName =
+            action.actionCustomizationArgs.state_name.value;
+      }
+      for (i = indexOfFirstDisplayedAction - 1; i >= 0; i--) {
+        let action = playthrough.actions[i];
+        let currentStateName =
+            action.actionCustomizationArgs.state_name.value;
+
+        if (currentStateName !== previousStateName) {
+          // When the latest state name and the state name of the
+          // current learner action differ, they are part of
+          // different contexts. So, we identify whether the size
+          // of the block has crossed the threshold number of actions
+          // per block. If it has, we don't add more learner actions
+          // to the block.
+          if (indexOfFirstDisplayedAction - i >= (
```

Comment:
Remove second set of parenthesis and indent next line.

Reply:
Done.